### ISSUE TITLE
UserVars.NutOnBatteryDelay parameter issue

### ISSUE THREAD

**Issue by YuriyDY** [rgc2000/NutClient-ESXi#5]
Dear Rene,

The version of my ESXi host is ESXi-7.0U3d-19482537

I am installed actual NutClient-ESXi module and its working fine with the default timers: 
![Untitled](https://github.com/rgc2000/NutClient-ESXi/assets/77094592/d6107894-3d8e-417f-bf23-713b139f4120)
After NUT notified the ESXi host of the low battery, it immediately shut down.

I tryed to change the UserVars.NutOnBatteryDelay parameter to 10 seconds and restarted the server:
![Untitled1](https://github.com/rgc2000/NutClient-ESXi/assets/77094592/cb0b9d30-e0bb-44e0-9b7a-c27145da4ca3)

Everething is fine, NUT client was runned:
![image](https://github.com/rgc2000/NutClient-ESXi/assets/77094592/b7cb05b5-70df-44f1-8c8f-d27011a5b52b)

And then I started the test. 
![image](https://github.com/rgc2000/NutClient-ESXi/assets/77094592/bbbc7686-61e2-408d-9338-55fd38437f14)
![image](https://github.com/rgc2000/NutClient-ESXi/assets/77094592/105a5ebe-cd9a-4386-8370-82e9f81e930f)

After 55 seconds ESXi continued to work:
![image](https://github.com/rgc2000/NutClient-ESXi/assets/77094592/0f4d4adf-dda8-4b82-b823-ef7423ff2d79)

Of course, while I was running the test, all the VMs were down. And I expected the host to shut down immediately after 10 seconds. But it still worked. 

Do you think that I missed something? Or this is a bug?

**Comment by rgc2000** [comment]
Have you restarted the upsmon NutClient service on ESXi after changing your settings? Otherwise, it does not take into account hot modifications.

**Comment by YuriyDY** [comment]
Yes, I tried to restart the NUT client service and the esxi host. The same issue

**Comment by rgc2000** [comment]
I have made a test on ESXi 7.0U3n (higher revision makes no difference) with NUT-Client-2.8.1-2.6.1 and the same parameters as you. I also sent OB DISCHRG from Nut Server. It works as expected. In ESXi /var/log/syslog.log I can see the power down decision exactly 10 seconds after the onbatt event :

```
2024-02-18T14:38:40.367Z upsmon[135154]: UPS test1@upserver on battery
2024-02-18T14:38:50.385Z upsmon[135154]: Signal 10: User requested FSD
2024-02-18T14:38:50.385Z upsmon[135154]: Executing automatic power-fail shutdown
2024-02-18T14:38:50.388Z upsmon[135154]: Auto logout and shutdown proceeding

```

Can you tell me which version of the NUT-Client you are using ?

**Comment by YuriyDY** [comment]
I have upgraded my ESXi to 7.0U3n to be syncronized with you. And also I using latest NUT-Client-2.8.1-2.6.1. I installed it from NutClient-ESXi-2.8.1-2.6.1-offline_bundle.zip.

This is part of my log file:
```
2024-02-18T19:48:18.574Z NUT[2102046]: NUT client is running
2024-02-18T19:48:18.671Z sfcbd-init[2102067]: args ('status')
2024-02-18T19:48:18.677Z sfcbd-init[2102067]: Getting Exclusive access, please wait...
2024-02-18T19:48:18.737Z sfcbd-init[2102067]: Exclusive access granted.
2024-02-18T19:50:26.482Z upsmon[2102036]: UPS ippon1200@192.168.1.37 on battery
2024-02-18T19:51:46.529Z upsmon[2102036]: UPS ippon1200@192.168.1.37 on line power

```

May be I can provide you something else?

**Comment by rgc2000** [comment]
Strange. Can you post here the contents of ESXi file /opt/nut/etc/notify.conf ? You can anonymize personal data.
This file is created when you start the NutClient service.

**Comment by YuriyDY** [comment]
Agree. 
From the /opt/nut/etc/notify.conf everething looking good:
```
# notification configuration
UPS_LIST="ippon1200@192.168.1.37"
MINSUPPLIES="1"
ONBATT_DELAY="5" ## I try different values for this timer
SEND_MAIL="0"
SMTP_RELAY="none"
[ "${SMTP_RELAY}" = "none" ] && SMTP_RELAY=""
TO="root@domain"

```

I tryed to modify /opt/nut/bin/notify.sh with echo markers:
```
# count how many UPSes are still online
NB_UPS_ONLINE=$(for UPS in ${UPS_LIST}; do /opt/nut/bin/upsc "${UPS}" ups.status; done | grep -c OL)

# Delayed shutdown if running on battery with less than minsupplies online
echo "Start"
if [ "${NOTIFYTYPE}" = "ONBATT" -a "${NB_UPS_ONLINE}" -lt "${MINSUPPLIES}" -a "${ONBATT_DELAY}" -gt 0 ]
echo "Getting UPS nums"
then
  if [ ! -f "${SHUDOWN_PID_FILE}" ]
  then
    (
      # seconds to wait
      sleep "${ONBATT_DELAY}"
      echo "End timer"
      # force shutdown
      rm "${SHUDOWN_PID_FILE}"
      /opt/nut/sbin/upsmon -c fsd
      exit 0
    ) &
    echo $! > "${SHUDOWN_PID_FILE}"
  fi
fi

```
But I cannot see any echos in a command line or syslog file. Its looking like NOTIFY event cannot call notify.sh. But I cannot see any troubles here:

```
[root@esxi:~] cat /opt/nut/etc/upsmon.conf
# Network UPS Tools: upsmon configuration

SHUTDOWNCMD "poweroff"
**NOTIFYCMD "/opt/nut/bin/notify.sh"**

POLLFREQ 5
POLLFREQALERT 5
HOSTSYNC 15
DEADTIME 15

NOTIFYFLAG ONLINE   SYSLOG+EXEC
**NOTIFYFLAG ONBATT   SYSLOG+EXEC**

```

Can you please tell me paths to additional log files for analysis?

**Comment by rgc2000** [comment]
How have you ben able to modify the /opt/nut/bin/notify.sh file ? It is a read-only and protected file. Even root cannot modify it. You need to make changes in source files and recompile the package.

If you want to output a message to the syslog don't use echo but use command :
`
logger -t NUT "message"
`
the "**" are not part of the file upsmon.conf, right ? I suppose it is a markdown attempt to emphasize the text already in code format.

You can try a larger delay (120 or more) and check that you have a non empty file /var/run/ups_shutdown.pid after receiving the ONBATT event.

then execute this command to check that the shutdown process is running and executing a sleep command.
`ps -cv | grep $(cat /var/run/ups_shutdown.pid)`

To get the process tree use this command and find the process with the PID stored in /var/run/ups_shutdown.pid, you should see the sleep with the delay 
`ps -cJ`

Also what is the result of this command while UPS is on battery ? 
`/opt/nut/bin/upsc ippon1200@192.168.1.37 ups.status`

Last thing, give me the output of
`ls -l /opt/nut/bin`

**Comment by YuriyDY** [comment]
Yes right. I tryed to make changes to the source files, then do "make". And than I installed modified addon.

Now I've rolled it back. I will modify notify.sh with "logger" command if you want, but I found some unexpected behavior.
Yes, you are right. The "**" is the markdown attemt. Excuse me. 

I changed the battery delay timer to 120 seconds
![image](https://github.com/rgc2000/NutClient-ESXi/assets/77094592/0a960472-9d46-4388-8c27-30883c79e999)

And then I started the test:
![image](https://github.com/rgc2000/NutClient-ESXi/assets/77094592/fc798e03-d669-4633-ab37-f83da58bc11e)

The output for 
```
ls /var/run/ups_shutdown.pid
ps -cv | grep $(cat /var/run/ups_shutdown.pid)
ps -cJ | grep sleep
```
![image](https://github.com/rgc2000/NutClient-ESXi/assets/77094592/98b61287-e54a-4d71-be36-e696ec582169)

After end of the sleep timer:
![image](https://github.com/rgc2000/NutClient-ESXi/assets/77094592/6c7261d9-e20f-429b-9da7-bed6a20bdc97)
![image](https://github.com/rgc2000/NutClient-ESXi/assets/77094592/42286838-2352-4597-87f1-5e01daec84d2)

UPS status and bin folder looking good:
![image](https://github.com/rgc2000/NutClient-ESXi/assets/77094592/2ad9d925-84e6-43a6-a903-6266033b6180)

So, NUT CLient does't created ups_shutdown.pid. Is there a problem with my NUT Client installation?

**Comment by rgc2000** [comment]
One test is missing : the result of this command while UPS is on battery :
`/opt/nut/bin/upsc ippon1200@192.168.1.37 ups.status`

The other results are correct. The ups_shutdown.pid only exists while the process is waiting for the end of the ONBATT_DELAY. And just after deleting the file it should start the poweroff sequence. Therefore it uses a manual NUT event to force shutdown and it seems that this has no effect.

Could you try to send this event on command line ? This should immediately start a shutdown so your ESXi will poweroff : 
`/opt/nut/sbin/upsmon -c fsd`

If your ESXi ignores this command we'll have to search in the ESXi configuration. Can you confirm that your ESXi is not part of a VMware cluster with HA enabled ?

**Comment by YuriyDY** [comment]
You can find ups.status request and answer in a last screenshot. Everething is fine here:
![image](https://github.com/rgc2000/NutClient-ESXi/assets/77094592/d1582687-510a-4e39-a9b4-7439d636773e)

This is reaction to the upsmon command:
![image](https://github.com/rgc2000/NutClient-ESXi/assets/77094592/a2579800-1d61-487c-8765-42d2e91eb621)

Does it look like the solution to the problem is close?

Yes, my server was deploed in a standalone install without HA and vSphere:
![image](https://github.com/rgc2000/NutClient-ESXi/assets/77094592/b595df1e-bca2-437d-bcec-699b6a4ed2ff)

**Comment by rgc2000** [comment]
This is very strange, the location of the pid file for upsmon is not /run/upsmon.pid but /var/run/upsmon.pid
There is something wrong with your binary. The compilation process is not ok and the default has not been set as expected.

I can see that the configure script will use the default for the system where you are compiling. So the linux where you made the compilation is using /run instead of /var/run for privileged pid files.

Let's force it to the desired value. Can you edit the Makefile and change the line 42 to add the option **--with-pidpath=/var/run** to explicitly set the path to the pid file : 
`cd nut-$(NUT_VERSION); touch .git; ./configure --prefix=/opt/nut --with-pidpath=/var/run --without-cgi --without-snmp --without-wrap --without-serial --with-user=daemon --with-group=daemon --with-openssl --with-openssl-includes="-I$(CURDIR)/libressl-bin/inc        lude" --with-openssl-libs="-L$(CURDIR)/libressl-bin/lib -lssl -lcrypto" LDFLAGS="-lrt -lpthread"`

Then recompile the package and update/reinstall the package.

**Comment by YuriyDY** [comment]
Thank you so much! Now its working
![image](https://github.com/rgc2000/NutClient-ESXi/assets/77094592/42d65fb0-fec9-4166-b491-a60e9d183eca)

I used Centos 7 minimal to build the package. 
![image](https://github.com/rgc2000/NutClient-ESXi/assets/77094592/2a3a3b42-a84f-4feb-841d-5a3ceb7455c1)

Can I provide something else for you about my Centos installation?

**Comment by rgc2000** [comment]
I am still using an old CentOS 6 to compile the package. I have tested that the compilation works under CentOS 7 and Linux Rocky 8 but I have not tested that the binaries were working as expected.

I am going to add this change to the Makefile in git repository but as this is only a build issue I will not release a new version of the package. The provided binaries are OK.

Thank you for your feedback. This will make the code more portable on other Linux distributions for users who want to edit and compile the sources.

**Comment by rgc2000** [comment]
Build with /var/run path for upsmon pid file fixed in commit ce40ce618f86f802e234324b3887108299a11523
